src/Npgsql/NpgsqlParameter.cs

```diff
@@ -368,7 +368,22 @@ public NpgsqlDbType NpgsqlDbType
             {
                 if (_dataTypeName != null)
                     return _dataTypeName;
-                throw new NotImplementedException("Infer from others");
+                if (_npgsqlDbType.HasValue)
+                {
+                    var primitiveNpgsqlType = _npgsqlDbType.Value & ~NpgsqlDbType.Array & ~NpgsqlDbType.Range;
+                    var primitiveTypeName = GlobalTypeMapper.Instance.ToPgTypeName(primitiveNpgsqlType);
+
+                    if ((_npgsqlDbType.Value & NpgsqlDbType.Array) != 0 ||
+                        (_npgsqlDbType.Value & NpgsqlDbType.Range) != 0)
+                    {
+                        primitiveTypeName += "[]";
+                    }
+
+                    return primitiveTypeName;
+                }
+                if (_value != null)   // Infer from value
+                    return GlobalTypeMapper.Instance.ToPgTypeName(_value.GetType());
+                throw new InvalidOperationException($"Parameter '{ParameterName}' type not set and no value was supplied");
```

Comment:
Am not sure what `ParameterName` has to do with inferring the DataTypeName. Also, when we can't infer, we generally return Unknown or similar (you also return null above) - so is there a reason to throw here?

Reply:
My idea was to use `ParameterName` to provide some context for this exception. Anyhow following you suggestion I changed the code to return null whenever the TypeName cannot be inferred.